File: apps/api/src/tokens.ts
/**
 * Access tokens are RS256 and short (10 minutes) so a verifier never holds a
 * signing key and a stolen token expires on its own. Refresh tokens are opaque,
 * stored only as a hash, rotated on every use, and a replayed one kills the
 * whole session family.
 */

import fs from 'node:fs/promises'
import path from 'node:path'
import { randomBytes } from 'node:crypto'
import { SignJWT, jwtVerify, importPKCS8, importSPKI, exportPKCS8, exportSPKI, generateKeyPair } from 'jose'
import type { KeyLike } from 'jose'
import { env } from './env.ts'

const ALG = 'RS256'
const ISS = 'brolly'
const AUD = 'brolly-api'

type KeyPair = { priv: KeyLike; pub: KeyLike }
let keys: KeyPair | null = null

async function loadKeys(): Promise<KeyPair> {
  if (keys) return keys
  const privPath = path.join(env.keyDir, 'access.key')
  const pubPath = path.join(env.keyDir, 'access.pub')
  let pair: KeyPair
  try {
    const [privPem, pubPem] = await Promise.all([fs.readFile(privPath, 'utf8'), fs.readFile(pubPath, 'utf8')])
    pair = { priv: await importPKCS8(privPem, ALG), pub: await importSPKI(pubPem, ALG) }
  } catch {
    // First boot: mint a keypair and keep it, so restarting the API does not
    // sign every user out.
    const kp = await generateKeyPair(ALG, { extractable: true })
    await fs.mkdir(env.keyDir, { recursive: true })
    await fs.writeFile(privPath, await exportPKCS8(kp.privateKey), { mode: 0o600 })
    await fs.writeFile(pubPath, await exportSPKI(kp.publicKey))
    pair = { priv: kp.privateKey, pub: kp.publicKey }
  }
  keys = pair
  return pair
}

export type AccessClaims = {
  sub: string
  tid: string
  scope: 'tenant' | 'platform'
  sid: string
  pv: number
  roles: string[]
}

export async function signAccessToken(c: AccessClaims): Promise<string> {
  const { priv } = await loadKeys()
  return new SignJWT({ tid: c.tid, scope: c.scope, sid: c.sid, pv: c.pv, roles: c.roles })
    .setProtectedHeader({ alg: ALG, kid: 'access-1' })
    .setSubject(c.sub)
    .setIssuer(ISS)
    .setAudience(AUD)
    .setIssuedAt()
    .setExpirationTime(env.accessTokenTtl)
    .sign(priv)
}

export async function verifyAccessToken(token: string): Promise<AccessClaims> {
  const { pub } = await loadKeys()
  const { payload } = await jwtVerify(token, pub, { issuer: ISS, audience: AUD, algorithms: [ALG] })
  return {
    sub: String(payload.sub),
    tid: String(payload.tid),
    scope: payload.scope as 'tenant' | 'platform',
    sid: String(payload.sid),
    pv: Number(payload.pv),
    roles: (payload.roles as string[]) ?? [],
  }
}

export const newRefreshToken = () => randomBytes(32).toString('base64url')

File: backend/app/core/tokens.py
"""
Access tokens are RS256 and short (10 minutes) so a verifier never holds a
signing key and a stolen token expires on its own. Refresh tokens are opaque,
stored only as a hash, rotated on every use, and a replayed one kills the whole
session family.
"""
from __future__ import annotations

import secrets
import time
from dataclasses import dataclass
from typing import Literal

import jwt
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import rsa

from ..config import settings

ALG = "RS256"
ISS = "brolly-juniors"
AUD = "brolly-b2c-api"

_private_pem: bytes | None = None
_public_pem: bytes | None = None


def _load_keys() -> tuple[bytes, bytes]:
    global _private_pem, _public_pem
    if _private_pem is not None and _public_pem is not None:
        return _private_pem, _public_pem

    key_dir = settings.key_dir
    priv_path = key_dir / "access.key"
    pub_path = key_dir / "access.pub"

    if priv_path.exists() and pub_path.exists():
        _private_pem = priv_path.read_bytes()
        _public_pem = pub_path.read_bytes()
        return _private_pem, _public_pem

    # First boot: mint a keypair and keep it, so restarting the API does not
    # sign every user out.
    key = rsa.generate_private_key(public_exponent=65537, key_size=2048)
    priv = key.private_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PrivateFormat.PKCS8,
        encryption_algorithm=serialization.NoEncryption(),
    )
    pub = key.public_key().public_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PublicFormat.SubjectPublicKeyInfo,
    )
    key_dir.mkdir(parents=True, exist_ok=True)
    priv_path.write_bytes(priv)
    pub_path.write_bytes(pub)
    try:
        priv_path.chmod(0o600)
    except OSError:
        pass  # Windows ACLs; the directory is already user-scoped

    _private_pem, _public_pem = priv, pub
    return priv, pub


@dataclass(slots=True)
class AccessClaims:
    sub: str
    #: Primary role. There is no tenant claim, because there are no tenants.
    role: Literal["BROLLY_ADMIN", "TEACHER", "STUDENT"]
    sid: str
    pv: int


def sign_access_token(claims: AccessClaims) -> str:
    priv, _ = _load_keys()
    now = int(time.time())
    return jwt.encode(
        {
            "sub": claims.sub,
            "role": claims.role,
            "sid": claims.sid,
            "pv": claims.pv,
            "iss": ISS,
            "aud": AUD,
            "iat": now,
            "exp": now + settings.access_token_ttl_seconds,
        },
        priv,
        algorithm=ALG,
        headers={"kid": "access-1"},
    )


def verify_access_token(token: str) -> AccessClaims:
    _, pub = _load_keys()
    payload = jwt.decode(
        token, pub, algorithms=[ALG], issuer=ISS, audience=AUD,
    )
    return AccessClaims(
        sub=str(payload["sub"]),
        role=payload["role"],
        sid=str(payload["sid"]),
        pv=int(payload["pv"]),
    )


def new_refresh_token() -> str:
    return secrets.token_urlsafe(32)
